src/cephadm/cephadm

```diff
@@ -5452,6 +5548,593 @@ def command_gather_facts():
+            elif u == 'daemons':
+                data = json.dumps(self.server.cephadm_cache.daemons)
+            elif u == 'disks':
+                data = json.dumps(self.server.cephadm_cache.disks)    
+            elif u == 'health':
+                data = json.dumps(self.server.cephadm_cache.health)
+            elif u == 'host':
+                data = json.dumps(self.server.cephadm_cache.host)
+            
+            self.send_response(200)
+            self.send_header('Content-type','application/json')
+            self.end_headers()
+            self.wfile.write(data.encode('utf-8'))
+        else:
+            # Invalid GET URL
+            bad_request_msg = "Valid URLs are: {}".format(', '.join(CephadmDaemonHandler.valid_routes))
+            self.send_response(404, message=bad_request_msg)  # reason
+            self.send_header('Content-type','application/json')
+            self.end_headers()
+            self.wfile.write(json.dumps({"message": bad_request_msg}).encode('utf-8'))
+    
+    def log_message(self, format, *args):
+        rqst = " ".join(str(a) for a in args)
+        logger.info(f"client:{self.address_string()} [{self.log_date_time_string()}] {rqst}")
+
+
+class CephadmDaemon():
+
+    daemon_type = "cephadm-exporter"
+    default_port = 9443
+    bin_name = 'cephadm'
+    key_name = "key"
+    crt_name = "crt"
+    token_name = "token"
+    config_requirements = [
+        key_name,
+        crt_name,
+        token_name,
+    ]
+    loop_delay = 1
+    thread_check_interval = 5
+
+    def __init__(self, fsid, daemon_id=None, port=None):
+        self.fsid = fsid
+        self.daemon_id = daemon_id 
+        if not port:
+            self.port = CephadmDaemon.default_port
+        else:
+            self.port = port
+        self.workers = []
+        self.http_server: CephadmHTTPServer
+        self.stop = False
+        self.cephadm_cache = CephadmCache()
+        self.errors = []
+        self.token = read_file([os.path.join(self.daemon_path, CephadmDaemon.token_name)])
+
+    @classmethod
+    def validate_config(cls, config):
+        reqs = ", ".join(CephadmDaemon.config_requirements)
+        errors = []
+
+        if not config or not all(k_name in config for k_name in CephadmDaemon.config_requirements):
+            raise Error("config must contain the following fields : {}".format(reqs))
+        
+        if not isinstance(config[CephadmDaemon.crt_name], str) or not isinstance(config[CephadmDaemon.key_name], str):
+            errors.append("crt and key fields must be 'str' types")
+
+        token = config[CephadmDaemon.token_name]
+        if len(token) < 8 or not isinstance(token, str):
+            errors.append("token must be of type 'str' and more than 8 characters long")
+
+        if 'port' in config:
+            try:
+                p = int(config['port'])
+                if p <= 1024:
+                    raise ValueError
```

Comment:
I'm confused...the validation traps the exceptions to assemble an errors 'list'. If the list contains entries all the issues are listed with a single raise call - so you see all the errors at once instead of repeat invocations resolving each exception one by one.